marketplace_standard_app_api/main.py

```diff
@@ -108,29 +108,155 @@ async def heartbeat() -> HTMLResponse:
+    response_model=DatasetCreateResponse,
+    status_code=201,
+    responses={
+        401: {"description": "Not authenticated."},
+        500: {"description": "Internal server error."},
+        501: {"description": "Not implemented."},
+        503: {"description": "Service unavailable."},
+    },
+    description="Create or replace a dataset.\n" + CREATE_DATASET_DESCRIPTION,
+)
+@api.put(
+    "/datasets/",
     operation_id="createDataset",
     tags=["DataSink"],
     response_model=DatasetCreateResponse,
+    status_code=201,
     responses={
         401: {"description": "Not authenticated."},
         500: {"description": "Internal server error."},
         501: {"description": "Not implemented."},
         503: {"description": "Service unavailable."},
     },
+    description="Create a dataset.\n" + CREATE_DATASET_DESCRIPTION,
 )
-async def create_dataset(dataset: DatasetModel) -> DatasetCreateResponse:
-    """Create a new dataset."""
+async def create_dataset(
+    file: UploadFile,
+    dataset_id: Optional[DatasetId] = None,
+) -> DatasetCreateResponse:
+    """Create a new or replace an existing dataset."""
+    raise HTTPException(status_code=501, detail="Not implemented.")
+
+
+@api.post(
+    "/datasets/",
+    operation_id="createDatasetMetadata",
+    name="Create Dataset Metadata",
+    tags=["DataSink"],
+    status_code=202,
+    responses={
+        401: {"description": "Not authenticated."},
+        500: {"description": "Internal server error."},
+        501: {"description": "Not implemented."},
+        503: {"description": "Service unavailable."},
+    },
+)
+@api.post(
+    "datasets/{dataset_id}",
+    operation_id="createOrReplaceDatasetMetadata",
+    name="Create or Replace Dataset Metadata",
+    tags=["DataSink"],
+    status_code=202,
+    responses={
+        401: {"description": "Not authenticated."},
+        500: {"description": "Internal server error."},
+        501: {"description": "Not implemented."},
+        503: {"description": "Service unavailable."},
+    },
+)
+async def create_or_replace_dataset_metadata(
+    dataset_id: Optional[DatasetId] = None,
+):
+    """Create or replace dataset metadata.
+
+    Note: This operation is in compliance with the OpenStack Swift object
+    storage API:
+    https://docs.openstack.org/api-ref/object-store/index.html#create-or-update-object-metadata
+    """
+    raise HTTPException(status_code=501, detail="Not implemented.")
+
+
+@api.head(
+    "/datasets/{dataset_id}",
+    operation_id="getDatasetMetadata",
+    tags=["DataSource"],
+    status_code=200,
+    response_class=Response,
+    responses={
+        401: {"description": "Not authenticated."},
+        404: {"description": "Not found."},
+        500: {"description": "Internal server error."},
+        501: {"description": "Not implemented."},
+        503: {"description": "Service unavailable."},
+    },
+)
+async def get_dataset_metadata(dataset_id: DatasetId) -> Response:
```

Comment:
Previously we considered DCAT as means to describe metadata of datasets. DCAT is great since it describes well metadata that is closely related to datasets (i.e., creator, license, etc.). Here the metadata seems more generic, which could possibly also be sufficient but perhaps not as descriptive.

Reply:
Since there is no specification with respect to what kind of metadata can or must be supplied, I think we could introduce that later if you like.

Reply:
I think you would need the DCAT payload as a response body, but you can't send a response body with the HEAD method.
I would change nothing however, since I'm not sure how important it is.

Reply:
Hmmm.. not sure this is true, but I in that case I would suggest to use a two-fold approach. Upload the metadata as a separate dataset and then link them via the metadata header attributes. So our dataset would have a metadata key like this:
```
X-Object-Meta-dcat thiscollection/abcdef
```
I am going to keep this conversation resolved since it would be something to tackle at a later stage anyways. Thank  you very much for the suggestion!

Reply:
Thanks for the suggestion. It seems I was indeed wrong about the HEAD method, which makes things simpler. From https://www.rfc-editor.org/rfc/rfc9110.html#section-9.3.2:

> The HEAD method is identical to GET except that the server MUST NOT send content in the response.

Reply:
Yes, I am just not sure that it is not feasible to send the DCAT object as part of the header just like any other metadata. Is this part of the DCAT specification that it must be sent in the response content body? Again, if that's the case we could associate it with another object like I suggested above.

Reply:
Ah, I see. I haven't consider sending it as a header. Then maybe that could also be a solution.